ADBDEV-4726-37 Check that plan is not null

dnskvlnk opened the pull request:
Check that plan is not null

create_hash_join passes outer_plan to disuse_physical_tlist where it's
dereferenced without checking that it's not null. We can't guarantee that
outer_plan is always valid. This patch adds an assertion

Commit:
Check that plan is not null

Commit:
Check that plan is not null

disuse_physical_tlist dereferences plan without checking that it's not null.
This patch adds an assert

RekGRpth commented on src/backend/optimizer/plan/createplan.c:
```diff
@@ -640,6 +640,8 @@ use_physical_tlist(PlannerInfo *root, RelOptInfo *rel)
 static void
 disuse_physical_tlist(PlannerInfo *root, Plan *plan, Path *path)
 {
+	Assert(plan);
```
Maybe it would be better to add a check before calling the disuse_physical_tlist function, as was done before its other call in the create_hashjoin_plan function?

dnskvlnk replied:
I think it's better to leave it here so that we won't have to check a plan in every calling function

dnskvlnk replied:
Anyway, what's the sense to pass a null plan here?

dnskvlnk replied:
I decided to leave the check in disuse_physical_tlist because in your example there's no error in a case when outer_plan is null, it seems like it's expected and shouldn't be copied in other places

Commit:
Check that plan is not null

disuse_physical_tlist dereferences plan without checking that it's not null.
We can't guarantee that plan is always valid. This patch adds Insist

RekGRpth commented on src/backend/optimizer/plan/createplan.c:
```diff
@@ -640,6 +640,8 @@ use_physical_tlist(PlannerInfo *root, RelOptInfo *rel)
 static void
 disuse_physical_tlist(PlannerInfo *root, Plan *plan, Path *path)
 {
+	Insist(plan);
```
How about putting the call under a general condition instead?
https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/optimizer/plan/createplan.c#L3697-L3703
```c
	disuse_physical_tlist(root, inner_plan, best_path->jpath.innerjoinpath);
	if (outer_plan)
	{
		disuse_physical_tlist(root, outer_plan, best_path->jpath.outerjoinpath);

		/* If we expect batching, suppress excess columns in outer tuples too */
		if (best_path->num_batches > 1)
			disuse_physical_tlist(root, outer_plan, best_path->jpath.outerjoinpath);
	}
```
Or, conversely, replace the condition with an assertion?
```c
	disuse_physical_tlist(root, inner_plan, best_path->jpath.innerjoinpath);
	Assert (outer_plan != NULL)
	disuse_physical_tlist(root, outer_plan, best_path->jpath.outerjoinpath);

	/* If we expect batching, suppress excess columns in outer tuples too */
	if (best_path->num_batches > 1)
		disuse_physical_tlist(root, outer_plan, best_path->jpath.outerjoinpath);
```

dnskvlnk replied:
Are you sure that at this point outer_plan can't be NULL?  I understand that it's a hashjoin node but I can't say for sure that it wasn't intended

RekGRpth replied:
I think create_plan_recurse cannot return NULL

dnskvlnk replied:
Done

Commit:
Check that plan is not null

create_hash_join passes outer_plan to disuse_physical_tlist where it's
dereferenced without checking that it's not null. We can't guarantee that
outer_plan is always valid. This patch adds an assertion
